### PR TITLE
Feature 3.7/add locking instrumentation

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

This PR contains several improvements for cluster transactions and collection locking, which are somewhat related:

Enable larger follower transactions, by enabling intermediate commits on followers, too.
If a transaction is intermediate-committed on a follower but aborted on a leader, the follower will respond with a new error code `TRI_ERROR_CLUSTER_FOLLOWER_TRANSACTION_COMMIT_PERFORMED` (mapped to HTTP 409 Conflict), which the leader can use to drop the follower.

The PR also allows to lazily start a follower transaction on a follower from multiple leaders.
This can happen if we have 3 (or more) database servers and 2 shards, so that
- db server 1 is the leader for shard A
- db server 2 is the leader for shard B
- db server 3 is the follower for both shards A and B

In this case db server 1 may try to lazily start a follower transaction on db server 3 for shard A, and db server 2 may try to do the same for shard B. It is important that both operations succeed, because otherwise one of the calls would just drop db server 3 as a follower.
It is also important to note that each leader may only send information about "their" shards to the follower (i.e. either shard A or shard B), but may not send the entire list of shards used in the transaction. This happens at least for multi-collection AQL write queries that lazily start follower transactions for a write operation on a given shard.
The previous behavior for this case was that the follower responded with "transaction in use" when the second leader tried to lazily start the follower transaction, and that the follower was dropped.
With this PR, no followers should be dropped in this situation.

The PR adds a new test suite in file `tests/js/client/shell/shell-transaction-intermediate-commit-cluster.js` to test the changes. Some of the tests are only executed if there are at least 3 DB servers, so the PR also changes the number of DB servers for some of the tests to 3 (from previously 2).

Backport of https://github.com/arangodb/arangodb/pull/12654

This PR also adds metrics for collection locks:
- `arangodb_collection_lock_timeouts_exclusive`: Number of lock timeouts when trying to acquire collection exclusive locks 
- `arangodb_collection_lock_timeouts_write`: Number of lock timeouts when trying to acquire collection write locks 
- `arangodb_collection_lock_acquisition_micros`: Total amount of collection lock acquisition time [μs]
- `arangodb_collection_lock_acquisition_time`: Total collection lock acquisition time histogram [s]

And it reduces lock timeout on followers to 15 seconds.
Rationale: we should not have any locking conflicts on followers, generally. Any shard locking should be performed on leaders first, which will then, eventually replicate changes to followers. replication to followers is only done once the locks have been acquired on the leader(s). 

The latter two things are a backport of https://github.com/arangodb/arangodb/pull/12985

Enterprise companion PR: https://github.com/arangodb/enterprise/pull/584
Docs PR: https://github.com/arangodb/docs/pull/575

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] No backports required

### Testing & Verification

- [x] The behavior in this PR was *manually tested*
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new **integration tests** (e.g. in shell_client)

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12646/

### CODE REVIEW

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12654/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12656/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12657/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12664/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12667/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12672/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12674/

**Comment by jsteemann** [comment]
Tests blue

File: arangod/Transaction/Manager.cpp [hunk 1]
```diff
@@ -312,15 +316,71 @@ Result Manager::createManagedTrx(TRI_vocbase_t& vocbase, TRI_voc_tid_t tid,
     auto& buck = _transactions[bucket];
     auto it = buck._managed.find(tid);
     if (it != buck._managed.end()) {
+      if (ServerState::instance()->isDBServer() && isFollowerTransactionId(tid)) {
```
Comment (neunhoef) [inline arangod/Transaction/Manager.cpp:319]: Do we need a case here that we are looking at a legacy transaction id and yet this is a follower transaction? What if the transaction has the hint IS_FOLLOWER_TRANSACTION?

File: arangod/Transaction/Manager.cpp [hunk 2]
```diff
@@ -312,15 +316,71 @@ Result Manager::createManagedTrx(TRI_vocbase_t& vocbase, TRI_voc_tid_t tid,
...
-                       "' already used, (before creating).");
+                       std::string("transaction id ") + std::to_string(tid) +
+                       " already used, (before creating)");
     }
   }
 
   // enforce size limit per DBServer
-  options.maxTransactionSize =
-      std::min<size_t>(options.maxTransactionSize, Manager::maxTransactionSize);
+  if (ServerState::instance()->isDBServer() && isFollowerTransactionId(tid)) {
```
Comment (neunhoef) [inline arangod/Transaction/Manager.cpp:341]: See above.

File: arangod/Transaction/Manager.cpp [hunk 3]
```diff
@@ -720,9 +800,16 @@ Result Manager::updateTransaction(TRI_voc_tid_t tid, transaction::Status status,
     }
   } else {
     res = trx.abort();
-    if (res.ok() && wasExpired) {
+    if (intermediateCommits &&
+        ServerState::instance()->isDBServer() &&
+        isFollowerTransactionId(tid)) {
```
Comment (neunhoef) [inline arangod/Transaction/Manager.cpp:805]: See above.

**Review by neunhoef — CHANGES_REQUESTED** [review]
Generally: LGTM. I have added 3 comments in places, where I think we need to distinguish another case: The case that we are looking at a follower transaction which has a "legacy" transaction ID.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12737/

File: arangod/Transaction/Manager.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/Transaction/Manager.cpp]: Done

File: arangod/Transaction/Manager.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/Transaction/Manager.cpp]: Done

File: arangod/Transaction/Manager.cpp (hunk 3 above)
Comment (jsteemann) [reply arangod/Transaction/Manager.cpp]: Done

**Review by neunhoef — APPROVED** [review]
LGTM

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12750/

**Comment by jsteemann** [comment]
Yet another run, due to MacOS compile warning:
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12752/

**Comment by jsteemann** [comment]
Tests blue

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12987/

**Comment by KVS85** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/13228/

**Comment by KVS85** [comment]
Tests blue.

**Review by KVS85 — APPROVED** [review]
LGTM.

**Pull request merged**